bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java

```diff
@@ -272,7 +277,7 @@ private CompletableFuture<Void> removeFile(String scriptIdentifier) {
                 ScriptFileReference ref = scriptMap.remove(scriptIdentifier);
 
                 if (ref == null) {
-                    logger.warn("Failed to unload script '{}': script reference not found.", scriptIdentifier);
+                    logger.debug("Failed to unload script '{}': script reference not found.", scriptIdentifier);
```

Comment:
Why did you reduce this logger? If unloading fails it is very likely we create memory or thread leaks.

Reply:
The problem is that this would happen when some random, untracked files got deleted.
Example:
- /conf/automation/ruby/ is the main watchPath
- /conf/automation/ruby/a_file is excluded and not added to scriptMap (for whatever reason - e.g. it is excluded by AbstractScriptFileWatcher itself, or excluded by child class)
- that file got deleted, triggering processWatchEvent
- Now people see this warning

It happens consistently with Ruby because .gem subdir is excluded and during addon initialisation, some lock files were created and then deleted, triggering a watchEvent.

Secondly - I couldn't see a memory / thread leak risk here.

In all places that call `removeFile`, they checked whether the entry they're removing exists inside scriptMap first before calling it. One place this didn't happen is inside processWatchEvent for DELETE event - and that's exactly when this happened.

Would you prefer that a check is done in processWatchEvent too, and leave the WARN logging there? I'd be happy with that solution as well, although I think it's redundant. Are there any other reasons that scriptMap.remove would return null other than the fact that the given element didn't exist?

Reply:
Another update on this. Although adding the check to make sure that the given scriptIdentifier exists in scriptMap inside processWatchEvent DELETE does indeed avoid triggering that warning, I came across another case where that warning was erroneously issued despite it being perfectly fine:

During bundle termination (e.g. bundle:stop), removeFile was called twice: once by `factoryRemoved` and another time by `deactivate`. And because removeFile is async, the same scriptIdentifier gets included twice in the removal. The first one removed the file normally, unloading the scriptEngine and all. The second one would find that the scriptIdentifier no longer exists in scriptMap and then issued this warning.

Instead, I've changed the warning to be issued when the scriptIdentifier is still there when it shouldn't be.